sigmastate/src/main/scala/sigmastate/serialization/ContractTemplateSerializer.scala

```diff
@@ -0,0 +1,144 @@
+    * Doesn't apply any transformations to the values. */
+  def deserializeContractTemplate(bytes: Array[Byte]): ContractTemplate = {
+    val r = SigmaSerializer.startReader(bytes)
+    deserializeContractTemplate(r)
+  }
+
+  private[sigmastate] def deserializeContractTemplate(r: SigmaByteReader): ContractTemplate = {
+    val startPos = r.position
+    val treeVersion = r.getOption(r.getUByte().toByte)
+
+    val nameLength = r.getUInt().toInt
+    val name = new String(r.getBytes(nameLength), StandardCharsets.UTF_8)
+
+    val descriptionLength = r.getUInt().toInt
+    val description = new String(r.getBytes(descriptionLength), StandardCharsets.UTF_8)
+
+    val nConstants = r.getUInt().toInt
+    val constTypes: IndexedSeq[SType] = {
+      if (nConstants > 0) {
+        // HOTSPOT:: allocate new array only if it is not empty
+        val res = safeNewArray[SType](nConstants)
+        cfor(0)(_ < nConstants, _ + 1) { i =>
+          res(i) = TypeSerializer.deserialize(r)
+        }
+        res
+      } else {
+        SType.EmptySeq
+      }
+    }
+    val constValues: Option[IndexedSeq[Option[SType#WrappedType]]] = r.getOption((() => {
+      if (nConstants > 0) {
+        // HOTSPOT:: allocate new array only if it is not empty
+        val res = safeNewArray[Option[SType#WrappedType]](nConstants)
+        cfor(0)(_ < nConstants, _ + 1) { i =>
+          res(i) = r.getOption((() => DataSerializer.deserialize(constTypes(i), r))())
+        }
+        res
+      } else {
+        Array.empty[Option[SType#WrappedType]]
+      }
+    })())
+
+    val nParameters = r.getUInt().toInt
+    val parameters: IndexedSeq[Parameter] = {
+      if (nParameters > 0) {
+        val res = safeNewArray[Parameter](nParameters)
+        cfor(0)(_ < nParameters, _ + 1) { i =>
+          res(i) = ParameterSerializer.deserialize(r)
+        }
+        res
+      } else {
+        Parameter.EmptySeq
+      }
+    }
+
+    // TODO: Here we need to create the list of constants (IndexedSeq[Constant[T]]) and populate it in the constant
+    // store as r.constantStore so that the expressionTree which contains ConstantPlaceholder's can be deserialized.
+    // Unfortunately, the value part of the constant is not always known here and can be None here.
+//    constTypes.indices.map(i => DeserializationSigmaBuilder.mkConstant(null, constTypes(i)))
+
+    val expressionTreeSize = r.getUInt().toInt
+    val expressionTree = ValueSerializer.deserialize(r.getBytes(expressionTreeSize))
+    CheckDeserializedScriptIsSigmaProp(expressionTree)
+
+    val contractTemplateSize = r.position - startPos
+    r.position = startPos
+    val templateBytes = r.getBytes(contractTemplateSize)
+
+    ContractTemplate(
+      treeVersion, name, description,
+      constTypes, constValues, parameters,
+      Right(expressionTree.toSigmaProp), templateBytes)
+  }
+}
+
+object ContractTemplateSerializer {
+  val DefaultSerializer = new ContractTemplateSerializer
```

Comment:
use SigmaBoolean as an example, not ErgoTree.

Reply:
Done